### ISSUE TITLE
Adding MSVC-Runtime Feature to spdlog-build

### ISSUE THREAD

**Issue by HpLightcorner** [gabime/spdlog#1608]
Hey spdlog-team,

I am in love with spdlog - it is always the first library I add - so foremost: thanks for your work. 

For a current project, I need to compile all my libraries using MSVC-Static Runtime Library. As I did not find a solution in the current spdlog-build configuration, I figured out that modern CMake offers this feature. Hence I slightly edited spdlog CMake configuration. I am definitely not a CMake professional, so I am not sure if my approach will handle all spdlog builds/flavours. However, it may be a good idea to add the feature? Here are my changes:

```
# Copyright(c) 2019 spdlog authors Distributed under the MIT License (http://opensource.org/licenses/MIT)

cmake_minimum_required(VERSION 3.2)

if(${CMAKE_VERSION} VERSION_LESS 3.11)
    cmake_policy(VERSION ${CMAKE_MAJOR_VERSION}.${CMAKE_MINOR_VERSION})
else()
    cmake_policy(VERSION 3.11)
endif()

# enable MSVC_RUNTIME_LIBRARY target property
# see https://cmake.org/cmake/help/latest/policy/CMP0091.html
if(POLICY CMP0091)
   cmake_policy(SET CMP0091 NEW)
endif()

enable_language(C)

# ---------------------------------------------------------------------------------------
# Start spdlog project
# ---------------------------------------------------------------------------------------
include(cmake/utils.cmake)
include(cmake/ide.cmake)

spdlog_extract_version()

project(spdlog VERSION ${SPDLOG_VERSION} LANGUAGES CXX)
message(STATUS "Build spdlog: ${SPDLOG_VERSION}")

include(GNUInstallDirs)

# ---------------------------------------------------------------------------------------
# Set default build to release
# ---------------------------------------------------------------------------------------
if(NOT CMAKE_BUILD_TYPE)
    set(CMAKE_BUILD_TYPE "Release" CACHE STRING "Choose Release or Debug" FORCE)
endif()

# ---------------------------------------------------------------------------------------
# Set MSVC-Runtime Library
# ---------------------------------------------------------------------------------------
set(CMAKE_MSVC_RUNTIME_LIBRARY
    MultiThreaded$<$<CONFIG:Debug>:Debug>)
message("CMAKE_MSVC_RUNTIME_LIBRARY ${CMAKE_MSVC_RUNTIME_LIBRARY}")

....
```

Maybe there is a CMake professional around or I can also try to make everything work and start a pull request. Or maybe spdlog already offers the configuration and I have overseen it?
Kind regards,
Heinz

**Comment by tt4g** [comment]
IMO.

To allow the user to specify the MSVC runtime, it would be preferable to have a pull request submitted before the `project()` to enable `CMP0091`.

If `CMP0091` is enabled, the user should be able to specify `cmake` argument `-DCMAKE_MSVC_RUNTIME_LIBRARY="MultiThreaded$<$<CONFIG:Debug>:Debug>"`.

Example for build static runtime: `cmake . -DCMAKE_MSVC_RUNTIME_LIBRARY="MultiThreaded$<$<CONFIG:Debug>:Debug>"`

**Comment by HpLightcorner** [comment]
Yes, this is of course a good solution.
`CMP0091` must be enabled before `project()` and also before `enable_language()`...

In addition, I would also suggest to document how to use CMake this way... I mean, maybe this is an "advanced" feature and the user who needs this should already be aware of it - however, I think adding some documentation how to build with different runtime settings `cmake . -DCMAKE_MSVC_RUNTIME_LIBRARY="....."`, giving the users some hints on this topic, would be a good idea.

**Comment by arthur-tacca** [comment]
With one exception, spdlog doesn't actually build a library at all, it is completely header only. That exception is the fmt library it depends upon, which by default is bundled with spdlog. The easiest way to build fmt statically is not to fiddle around with spdlog's build system but instead to build fmt directly with the options you want, then turn off spdlog's option to bundle fmt automatically (set `SPDLOG_FMT_EXTERNAL=ON`).

An easy way to do all this is to use [vcpkg](https://github.com/microsoft/vcpkg). This command one builds fmt with static linkage and against the static runtime, and copies the build results alongside spdlog into an "install" directory:

    vcpkg install --triplet x64-windows-static spdlog

**Comment by arthur-tacca** [comment]
Hmm, weird - it seems that spdlog does actually compile some files and create a library even if fmt is not bundled. But the vcpkg portfile (which is the recipe for building a library) for spdlog just [deletes the library file](https://github.com/microsoft/vcpkg/blob/ffe8f5d9b7818c208058b0c3f6a795fba0707db5/ports/spdlog/portfile.cmake#L45-L47) before copying to the installation directory. I and many others have used spdlog via vcpkg, including on Windows with dynamic and static linkage, and mixed linkage (static third-party libraries with dynamic VS runtime), and on Linux (static linkage), with no problems. So I think the spdlog library is built but is not actually necessary for use.

In other words, if you need to use spdlog with static linkage, try building fmt as you wish and then just use the headers of spdlog without building an spdlog library at all. But I still recommend you try vcpkg to do this, especially if you have other third-party libraries to build and use.

**Comment by tt4g** [comment]
@arthur-tacca spdlog 1.6.0 adds build support for Windows DLL: https://github.com/gabime/spdlog/releases/tag/v1.6.0
and build static library support is available from version 1.4.0: https://github.com/gabime/spdlog/releases/tag/v1.4.0

**Comment by HpLightcorner** [comment]
I see - in the end it may be just a lack of knowledge on my side, learning CMake and Build-Systems is still a hard way - and diving into each library and the individual build system costs resources I do not have. In the end it always takes a lot of time to get everything build correctly... On the other hand it gets easier with each project.... I will close the issue!

Thanks for your support!!
